src/coreclr/src/vm/dllimport.cpp

```diff
@@ -3046,9 +3046,21 @@ BOOL NDirect::MarshalingRequired(MethodDesc *pMD, PCCOR_SIGNATURE pSig /*= NULL*
         if (GetLCIDParameterIndex(pMD) != -1)
             return TRUE;
 
-        // making sure that cctor has run may be handled by stub
-        if (pMD->IsNDirect() && ((NDirectMethodDesc *)pMD)->IsClassConstructorTriggeredByILStub())
-            return TRUE;
+        if (pMD->IsNDirect())
+        {
+            // A P/Invoke marked with UnmanagedCallersOnlyAttribute
+            // doesn't technically require marshalling. However, we
+            // don't support a DllImport with this attribute and we
+            // error out during IL Stub generation so we indicate that
+            // when checking if an IL Stub is needed.
+            if (pMD->HasUnmanagedCallersOnlyAttribute())
```

Comment:
The same check should be also in `IsMarshallingRequired(MethodDesc targetMethod)` in `src\coreclr\src\tools\aot\ILCompiler.ReadyToRun\Interop\IL\Marshaller.ReadyToRun.cs`